File: backend/app/schemas/community.py
"""Pydantic schemas for WAYN Community."""

from datetime import datetime
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel, Field, model_validator


class CommunityPostCreate(BaseModel):
    place_id: UUID
    text: str | None = Field(default=None, max_length=5000)
    image_url: str | None = Field(default=None, max_length=1024)
    rating: Decimal | None = Field(default=None)

    @model_validator(mode="after")
    def validate_content(self):
        if not self.text and not self.image_url:
            raise ValueError("Post must contain text or image")
        return self

    @model_validator(mode="after")
    def validate_rating(self):
        if self.rating is not None:
            if self.rating < Decimal("1.0") or self.rating > Decimal("5.0"):
                raise ValueError("Rating must be between 1.0 and 5.0")
            self.rating = round(self.rating, 1)
        return self


class CommunityPostUpdate(BaseModel):
    text: str | None = Field(default=None, max_length=5000)
    image_url: str | None = Field(default=None, max_length=1024)
    rating: Decimal | None = None

    @model_validator(mode="after")
    def validate_rating(self):
        if self.rating is not None:
            if self.rating < Decimal("1.0") or self.rating > Decimal("5.0"):
                raise ValueError("Rating must be between 1.0 and 5.0")
            self.rating = round(self.rating, 1)
        return self


class CommunityPostRead(BaseModel):
    id: UUID
    user_id: UUID
    place_id: UUID
    text: str | None
    image_url: str | None
    rating: Decimal | None
    is_visible: bool
    created_at: datetime
    updated_at: datetime

    author_name: str | None = None
    author_username: str | None = None
    author_avatar: str | None = None
    place_name: str | None = None

    likes_count: int = 0
    saves_count: int = 0
    comments_count: int = 0

    is_liked: bool = False
    is_saved: bool = False

    model_config = {
        "from_attributes": True,
    }


class CommunityCommentCreate(BaseModel):
    text: str = Field(
        min_length=1,
        max_length=2000,
    )


class CommunityCommentRead(BaseModel):
    id: UUID
    post_id: UUID
    user_id: UUID
    text: str
    is_visible: bool
    created_at: datetime
    updated_at: datetime

    author_name: str | None = None
    author_username: str | None = None
    author_avatar: str | None = None

    model_config = {
        "from_attributes": True,
    }

File: admin/src/types/community.ts
import type { PaginatedResponse } from './place'

/**
 * Mirrors `CommunityPostRead` (backend/app/schemas/community.py).
 * Viewer-specific fields (is_liked / is_saved / is_owner ...) are always
 * false in the admin context but kept to match the schema exactly.
 */
export type CommunityPostRead = {
  id: string
  user_id: string
  place_id: string
  text: string | null
  image_url: string | null
  rating: number | null
  is_visible: boolean
  created_at: string
  updated_at: string
  author_name: string | null
  author_username: string | null
  author_avatar: string | null
  place_name: string | null
  place_city: string | null
  author_points: number
  author_followers_count: number
  is_following_author: boolean
  is_owner: boolean
  likes_count: number
  saves_count: number
  comments_count: number
  is_liked: boolean
  is_saved: boolean
}

/** Mirrors `CommunityCommentRead`. */
export type CommunityCommentRead = {
  id: string
  post_id: string
  user_id: string
  text: string
  is_visible: boolean
  created_at: string
  updated_at: string
  author_name: string | null
  author_username: string | null
  author_avatar: string | null
}

export type CommunityPostsResponse = PaginatedResponse<CommunityPostRead>
export type CommunityCommentsResponse = PaginatedResponse<CommunityCommentRead>

export type CommunityPostsParams = {
  page: number
  limit: number
  search?: string
  is_visible?: boolean
}

/** Body of `PATCH /admin/community/posts/{id}/visibility` and comments equivalent. */
export type VisibilityUpdatePayload = {
  is_visible: boolean
}
